common/source/nrt_fpga_device.cpp

```diff
@@ -0,0 +1,183 @@
+#include "nrt_fpga_device.h"
+
+#include <thread>
+
+NRTFPGADevice::NRTFPGADevice() :
+	nrt_fpga_fd_(0),
+	nrt_fpga_thread_([&](){ this->Work(); }) {
+	CommonKeyPairs config = {
+		{"fpga_device_file", "/dev/iofpga"}
+	};
+	ReConfig(config);
+	if (ReConfig(config) != kNoError)
+		return;
+	nrt_fpga_thread_.Start();
+}
+
+NRTFPGADevice::~NRTFPGADevice() {
+	DeConfig();
+}
+
+error_condition	NRTFPGADevice::ReConfig(const CommonKeyPairs& config) {
+
+	auto it_device_file = config.find(kFPGADeviceFile);
+	if (it_device_file == config.end())
+		return make_error_condition(ErrorCode::kDeviceFileNotSpecified);
+
+	int fpga_device_fd = open(it_device_file->second.c_str(), O_RDWR | O_NONBLOCK);
+	if (fpga_device_fd < 0)
+		return make_error_condition(ErrorCode::kOpenFileError);
+
+	error_condition result = kNoError;
+	// Select configuration page, prepare to configure FPGA
+	ioctl(fpga_device_fd, IOFPGA_SEL_CONF, NULL);
+	// Disable FPGA
+	result = WriteIOFPGA(fpga_device_fd, 0x4, 0xaa550000, 4);
+	if (result != kNoError)
+		return result;
+	// Enable high speed capture input channel
+	result = WriteIOFPGA(fpga_device_fd, 0x20, 0x00ff, 2);
+	if (result != kNoError)
+		return result;
+	// Set capture AB pulse capture mode
+	result = WriteIOFPGA(fpga_device_fd, 0x90, 0xff, 1);
+	if (result != kNoError)
+		return result;
+	result = WriteIOFPGA(fpga_device_fd, 0x92, 0x0f, 1);
+	if (result != kNoError)
+		return result;
+	// Reconfigure FPGA
+	result = WriteIOFPGA(fpga_device_fd, 0, 0x55aa0001, 4);
+	if (result != kNoError)
+		return result;
+	// Enable FPGA
+	result = WriteIOFPGA(fpga_device_fd, 0x04, 0xaa550001, 4);
+	if (result != kNoError)
+		return result;
+
+	// Select data page
+	ioctl(fpga_device_fd, IOFPGA_SEL_DATA, NULL);
+	// TODO: Read FIFO entry count and Clear FIFO
+
+	// Initial gpio state
+	if (WriteIOFPGA(fpga_device_fd, kDOPData, do_pdata_, sizeof(do_pdata_)) != kNoError \
+		|| WriteIOFPGA(fpga_device_fd, kDOHData, do_pdata_, sizeof(do_pdata_)) != kNoError)
+		return result;
+
+	nrt_fpga_fd_ = fpga_device_fd;
+	return kNoError;
+}
+
+error_condition	NRTFPGADevice::DeConfig() {
+	close(nrt_fpga_fd_);
+	return kNoError;
+}
+
+error_condition NRTFPGADevice::Get(const string& key, vector<uint8_t>& result) {
+	static const error_condition kInvalidArgument(errc::invalid_argument);
+	if (key.empty())
+		return kInvalidArgument;
+	lock_guard<mutex> lock(nrt_fpga_mutex_);
+	if (key == kGPIOData) {
```

Comment:
Need to check the length of result to avoid segmentation fault. And we could read data straight from FPGA, no need to read from buffer.